pkg/server/openapiv3/controller.go

```diff
@@ -0,0 +1,260 @@
+	log.Info("Starting controller")
+	defer log.Info("Shutting down controller")
+
+	if !cache.WaitForCacheSync(ctx.Done(), c.crdsSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	crds, err := c.crdLister.List(labels.Everything())
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to initially list all CRDs: %v", err))
+		return
+	}
+	for _, crd := range crds {
+		c.updateCRD(crd)
+	}
+
+	go wait.Until(func() { c.startWorker(ctx) }, time.Second, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (c *Controller) startWorker(ctx context.Context) {
+	for c.processNextWorkItem(ctx) {
+	}
+}
+
+func (c *Controller) processNextWorkItem(ctx context.Context) bool {
+	// Wait until there is a new item in the working queue
+	k, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	key := k.(string)
+
+	log := logging.WithQueueKey(klog.FromContext(ctx), key)
+	ctx = klog.NewContext(ctx, log)
+	log.V(4).Info("processing key")
+
+	// No matter what, tell the queue we're done with this key, to unblock
+	// other workers.
+	defer c.queue.Done(key)
+
+	// log slow aggregations
+	start := time.Now()
+	defer func() {
+		elapsed := time.Since(start)
+		if elapsed > time.Second {
+			log.Info("slow openapi v3 aggregation", "duration", elapsed)
+		}
+	}()
+
+	if requeue, err := c.process(ctx, key); err != nil {
+		utilruntime.HandleError(fmt.Errorf("%q controller failed to sync %q, err: %w", ControllerName, key, err))
+		c.queue.AddRateLimited(key)
+		return true
+	} else if requeue {
+		// only requeue if we didn't error, but we still want to requeue
+		c.queue.Add(key)
+		return true
+	}
+	c.queue.Forget(key)
+	return true
+}
+
+func (c *Controller) process(ctx context.Context, key string) (bool, error) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	clusterName, _, name, err := kcpcache.SplitMetaClusterNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return false, nil
+	}
+	crd, err := c.crdLister.Cluster(clusterName).Get(name)
+	if err != nil && !errors.IsNotFound(err) {
+		return false, err
+	}
+
+	if errors.IsNotFound(err) || !apiextensionshelpers.IsCRDConditionTrue(crd, apiextensionsv1.Established) {
+		delete(c.byClusterNameVersion[clusterName], name)
+		if len(c.byClusterNameVersion[clusterName]) == 0 {
+			delete(c.byClusterNameVersion, clusterName)
+		}
+		return false, nil
+	}
+
+	c.updateCRD(crd)
+
+	return false, nil
+}
+
+func (c *Controller) updateCRD(crd *apiextensionsv1.CustomResourceDefinition) {
+	clusterName := logicalcluster.From(crd)
+
+	// remove old instance
+	delete(c.byClusterNameVersion[clusterName], crd.Name)
+
+	// add new instance with all versions
+	for _, v := range crd.Spec.Versions {
+		if !v.Served {
+			delete(c.byClusterNameVersion[clusterName][crd.Name], v.Name)
+			continue
+		}
+
+		spec := cached.NewStaticSource[*spec3.OpenAPI](func() cached.Result[*spec3.OpenAPI] {
+			spec, err := builder.BuildOpenAPIV3(crd, v.Name, builder.Options{V2: false})
```

Comment:
Just trying to learn k8s machienry code,  what does this do?

Reply:
build a full spec from the openapi v3 schema in the CRD, i.e. add everything like http paths, parameters etc.